src/rpc/blockchain.cpp

```diff
@@ -2836,6 +2836,15 @@ static RPCHelpMan loadtxoutset()
             strprintf("The base block header (%s) must appear in the headers chain. Make sure all headers are syncing, and call this RPC again.",
                       base_blockhash.ToString()));
     }
+
+    bool start_block_invalid = WITH_LOCK(::cs_main,
+            return snapshot_start_block->nStatus & BLOCK_FAILED_MASK);
+    if (start_block_invalid) {
```

Comment:
Some checks are within `ActivateSnapshot()`, others, like the added one, are in the rpc code (which means that they can't be covered in  unit tests for that function). Do you know if there is a reason for that?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30267#discussion_r1638706909

> Some checks are within `ActivateSnapshot()`, others, like the added one, are in the rpc code (which means that they can't be covered in unit tests for that function). Do you know if there is a reason for that?

I think it this is just a mistake, and it makes the checks less reliable because they are done without keeping cs_main locked. I suggested fixing this before in https://github.com/bitcoin/bitcoin/pull/27596#discussion_r1325193433.

Reply:
I think these were always separated like that. But sounds good to me to move them. I have added a refactoring commit which moves the other two early checks into `ActivateSnapshot()` and then I am adding the new check there as well.

Reply:
Looks good (except for the compiler issue with `Join()`) - I think it would be helpful to not only log the reason, but also return them with the RPC - for that, `ActivateSnapshot()` would need to return it, possibly as `util::Result` as suggested by ryanofsky, but that could be a follow-up.
Though for now, it might be helpful to expand the error ("Unable to load UTXO snapshot") to point the user to the debug log for the reason.

Reply:
Ah, yeah, I gave it a shot and it pretty simple so I am changing the return type as well in the refactoring commit now.